examples/phase_estimator.py

```diff
@@ -0,0 +1,182 @@
+"""Creates and simulates a phase estimator circuit.
+
+=== EXAMPLE OUTPUT ===
+Estimation with 2qbits.
+Actual, Estimation (Raw binary)
+0.0000, 0.0000 (00)
+0.1000, 0.0000 (00)
+0.2000, 0.2500 (01)
+0.3000, 0.2500 (01)
+0.4000, 0.5000 (10)
+0.5000, 0.5000 (10)
+0.6000, 0.5000 (10)
+0.7000, 0.7500 (11)
+0.8000, 0.7500 (11)
+0.9000, 0.0000 (00)
+RMS Error: 0.2915
+
+Estimation with 4qbits.
+Actual, Estimation (Raw binary)
+0.0000, 0.0000 (0000)
+0.1000, 0.1250 (0010)
+0.2000, 0.1875 (0011)
+0.3000, 0.3125 (0101)
+0.4000, 0.3750 (0110)
+0.5000, 0.5000 (1000)
+0.6000, 0.6250 (1010)
+0.7000, 0.6875 (1011)
+0.8000, 0.8125 (1101)
+0.9000, 0.8750 (1110)
+RMS Error: 0.0177
+
+Estimation with 8qbits.
+Actual, Estimation (Raw binary)
+0.0000, 0.0000 (00000000)
+0.1000, 0.1016 (00011010)
+0.2000, 0.1992 (00110011)
+0.3000, 0.3008 (01001101)
+0.4000, 0.3984 (01100110)
+0.5000, 0.5000 (10000000)
+0.6000, 0.6016 (10011010)
+0.7000, 0.6992 (10110011)
+0.8000, 0.8008 (11001101)
+0.9000, 0.8984 (11100110)
+RMS Error: 0.0011
+"""
+
+
+from collections import Counter
+import numpy as np
+
+import cirq
+from cirq.ops import ControlledGate
+from cirq.ops import SingleQubitMatrixGate
+
+
+def mode(lst):
+
+    """Return the most common element from a list.
+    """
+
+    data = Counter(lst)
+    return data.most_common(1)[0][0]
+
+
+def phase_op(phi):
+
+    """An example unitary 1-qubit operation U with an eigen vector |0>
+    and an eigen value exp(2*Pi*i*phi)
+    """
+
+    gate = SingleQubitMatrixGate(
+        matrix=np.array([[np.exp(2*np.pi*1.0j*phi), 0], [0, 1]]))
+    return gate
+
+
+class QftInverse(cirq.Gate, cirq.CompositeGate):
+
+    """Quantum gate for the inverse Quantum Fourier Transformation
+    """
+
+    def cr_inverse(self, k):
+
+        """A controlled unitary gate R_k that applies the following phase shift:
+        R_k|0> = |0>
+        R_k|1> = exp(-2 Pi i / 2^k)|1>
+        """
+
+        r_k = SingleQubitMatrixGate(
+            matrix=np.array([[1, 0], [0, np.exp(-2*np.pi*1.0j/2**k)]]))
+        return ControlledGate(r_k)
+
+
+    def default_decompose(self, qubits):
+
+        """A quantum circuit with the following recursive structure.
+        ---H----R_2---R_3---R_4--------------------------------------
+                 |     |     |
+        ---------@-----|-----|-----H----R_2---R_3--------------------
```

Comment:
This is not actually a recursive structure.

Reply:
fixed.